ArduCopter/mode_2dpos.cpp

```diff
@@ -0,0 +1,222 @@
+        loiter_nav->init_target();
+        attitude_control->input_thrust_vector_rate_heading(loiter_nav->get_thrust_vector(), target_yaw_rate, false);
+        break;
+
+    case AltHold_Landed_Ground_Idle:
+        attitude_control->reset_yaw_target_and_rate();
+        FALLTHROUGH;
+
+    case AltHold_Landed_Pre_Takeoff:
+        attitude_control->reset_rate_controller_I_terms_smoothly();
+        loiter_nav->init_target();
+        attitude_control->input_thrust_vector_rate_heading(loiter_nav->get_thrust_vector(), target_yaw_rate, false);
+        pos_control->relax_z_controller(0.0f);   // forces throttle output to decay to zero
+        break;
+
+    case AltHold_Takeoff:
+        // initiate take-off
+        if (!takeoff.running()) {
+            takeoff.start(constrain_float(g.pilot_takeoff_alt,0.0f,1000.0f));
+        }
+
+        // get avoidance adjusted climb rate
+        target_climb_rate = get_avoidance_adjusted_climbrate(target_climb_rate);
+
+        // set position controller targets adjusted for pilot input
+        takeoff.do_pilot_takeoff(target_climb_rate);
+
+        // run loiter controller
+        loiter_nav->update();
+
+        // call attitude controller
+        attitude_control->input_thrust_vector_rate_heading(loiter_nav->get_thrust_vector(), target_yaw_rate, false);
+        break;
+
+    case AltHold_Flying:
+        // set motors to full range
+        motors->set_desired_spool_state(AP_Motors::DesiredSpoolState::THROTTLE_UNLIMITED);
+
+#if PRECISION_LANDING == ENABLED
+        bool precision_loiter_old_state = _precision_loiter_active;
+        if (do_precision_loiter()) {
+            precision_loiter_xy();
+            _precision_loiter_active = true;
+        } else {
+            _precision_loiter_active = false;
+        }
+        if (precision_loiter_old_state && !_precision_loiter_active) {
+            // prec loiter was active, not any more, let's init again as user takes control
+            loiter_nav->init_target();
+        }
+        // run loiter controller if we are not doing prec loiter
+        if (!_precision_loiter_active) {
+            loiter_nav->update();
+        }
+#else
+        loiter_nav->update();
+#endif
+
+        // call attitude controller
+        attitude_control->input_thrust_vector_rate_heading(loiter_nav->get_thrust_vector(), target_yaw_rate, false);
+
+        // get avoidance adjusted climb rate
+        target_climb_rate = get_avoidance_adjusted_climbrate(target_climb_rate);
+
+        // update the vertical offset based on the surface measurement
+        copter.surface_tracking.update_surface_offset();
+
+        // Send the commanded climb rate to the position controller
+        // pos_control->set_pos_target_z_from_climb_rate_cm(target_climb_rate);
+        pos_control->use_z_control(false);
+        if (copter.rccar_flow_hold){
+            pos_control->update_throttle(0);
+        } else {
```

Comment:
If we are in flowhold then don't allow takeoff in throttle trainer. This may need to be removed for the whoop. If we are not in flowhold then the throttle is directly taken from the rc command